### PR TITLE
HIVE-24803: update metrics and allocation after kill trigger status becomes available

### PR DESCRIPTION
**Pull request by guptanikhil007**
### What changes were proposed in this pull request?

Update the metrics and resource allocations in the pool after kill trigger processing is completed.

### Why are the changes needed?

Following two bugs were found with respect to Azure HDInsight customers

1. Workload manger doesn't process allocation changes after KillTrigger which leads to existing queries not able to ffully utilize the resources of the query which is killed.  
2. It is also found that the Workload Manager thread is not updating the runningQueries metric for the pool  after kill trigger is processed making metrics unreliable afterwards.

These changes are done to fix the above two bugs.

### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?

Added a unit test to validate metrics before and after execution of kill trigger

### CODE REVIEW

**Comment by guptanikhil007** [comment]
Following PR requires review: @sankarh , @adesh-rao

File: ql/src/java/org/apache/hadoop/hive/ql/exec/tez/WorkloadManager.java [hunk 1]
```diff
@@ -744,6 +748,19 @@ private void processCurrentEvents(EventState e, WmThreadSyncWork syncWork) throw
     }
   }
 
+  private void updatePoolMetricsAfterKillTrigger(HashSet<String> poolsToRedistribute, KillQueryContext ctx) {
+    String poolName = ctx.getPoolName();
+    if (poolName != null) {
```
Comment (ashish-kumar-sharma) [inline ql/src/java/org/apache/hadoop/hive/ql/exec/tez/WorkloadManager.java:753]: use StringUtils.isEmpty(poolName)

File: ql/src/java/org/apache/hadoop/hive/ql/exec/tez/WorkloadManager.java [hunk 2]
```diff
@@ -2225,13 +2251,17 @@ private void resetAndQueueKill(Map<WmTezSession, KillQueryContext> toKillQuery,
     KillQueryContext killQueryContext, Map<WmTezSession, GetRequest> toReuse) {
 
     WmTezSession toKill = killQueryContext.session;
+    String poolName = toKill.getPoolName();
+    boolean validPoolName = poolName != null;
```
Comment (ashish-kumar-sharma) [inline ql/src/java/org/apache/hadoop/hive/ql/exec/tez/WorkloadManager.java:2255]: break this line into 2. so that code will be more readable.

**Review by ashish-kumar-sharma — CHANGES_REQUESTED** [review]

File: ql/src/java/org/apache/hadoop/hive/ql/exec/tez/WorkloadManager.java (hunk 1 above)
Comment (guptanikhil007) [reply ql/src/java/org/apache/hadoop/hive/ql/exec/tez/WorkloadManager.java:753]: using StringUtils.isNotBlank

File: ql/src/java/org/apache/hadoop/hive/ql/exec/tez/WorkloadManager.java (hunk 2 above)
Comment (guptanikhil007) [reply ql/src/java/org/apache/hadoop/hive/ql/exec/tez/WorkloadManager.java:2257]: Broke the code into two separate blocks

**Comment by guptanikhil007** [comment]
@sankarh @adesh-rao  @ashish-kumar-sharma 
Please Review the changes

File: itests/hive-unit/src/test/java/org/apache/hive/jdbc/TestWMMetricsWithTrigger.java [hunk 3]
```diff
@@ -0,0 +1,203 @@
...
+
+  private static WMTrigger wmTriggerFromTrigger(Trigger trigger) {
+    WMTrigger result = new WMTrigger("rp", trigger.getName());
+    result.setTriggerExpression(trigger.getExpression().toString());
+    result.setActionExpression(trigger.getAction().toString());
+    return result;
+  }
+
+  @Test(timeout = 30000)
+  public void testWmPoolMetricsAfterKillTrigger() throws Exception {
```
[the remark this answers is not in the document]
Comment (guptanikhil007) [reply itests/hive-unit/src/test/java/org/apache/hive/jdbc/TestWMMetricsWithTrigger.java:187]: Added verification step

File: ql/src/java/org/apache/hadoop/hive/ql/exec/tez/WorkloadManager.java [hunk 4]
```diff
@@ -677,6 +677,8 @@ private void processCurrentEvents(EventState e, WmThreadSyncWork syncWork) throw
           }
           wmEvent.endEvent(ctx.session);
         }
+        // Running query metrics needs to be updated for the pool
```
[the remark this answers is not in the document]
Comment (guptanikhil007) [reply ql/src/java/org/apache/hadoop/hive/ql/exec/tez/WorkloadManager.java:681]: done

File: ql/src/java/org/apache/hadoop/hive/ql/exec/tez/WorkloadManager.java [hunk 5]
```diff
@@ -744,6 +748,21 @@ private void processCurrentEvents(EventState e, WmThreadSyncWork syncWork) throw
     }
   }
 
+  private void updatePoolMetricsAfterKillTrigger(HashSet<String> poolsToRedistribute, KillQueryContext ctx) {
+    String poolName = ctx.getPoolName();
+    if (StringUtils.isNotBlank(poolName)) {
+      poolsToRedistribute.add(poolName);
+      PoolState pool = pools.get(poolName);
+      if (pool != null) {
+        if (pool.metrics != null) {
```
[the remark this answers is not in the document]
Comment (guptanikhil007) [reply ql/src/java/org/apache/hadoop/hive/ql/exec/tez/WorkloadManager.java]: done

**Comment by sankarh** [comment]
+1, LGTM

**Pull request merged**